File: python/toposort.py


Graph = []


Graph.append([1,3])
Graph.append([3])
Graph.append([1,3])
Graph.append([4])
Graph.append([])





def dfs(at, visited, visitedVerteces, Graph):
    visited[at] = 1
    for vertex in Graph[at]:
        if not visited[vertex]:
            dfs(vertex,visited, visitedVerteces, Graph)
    visitedVerteces.append(at)


def topsort(Graph):
    V = len(Graph)
    visited = [0] * V
    order = [0]*V
    i = V-1
    for at in visited:
        if not at:
            visitedVerteces = []
            dfs(at, visited, visitedVerteces, Graph)
            for vertex in visitedVerteces:
                order[i] = vertex
                i-=1


    return order

print(topsort(Graph))

File: cpp/toposort.cpp
#include<iostream>
#include<vector>
using namespace std;


bool visited[int(10e5)];

void dfs(int at, vector<int> visitedByDFS, bool visited[], int V,vector<int> Graph[]){

    visitedByDFS.push_back(at);
    visited[at] = 1;

    for(int next : Graph[at]){
        if(!visited[next]){
            dfs(next, visitedByDFS, visited, V, Graph);
        }
    }

}



int main(){


    int V, E;
    cin >> V >> E;
    vector<int> Graph[V];


    while(E){

        int a, b;
        cin >> a,b;
        a--;
        b--;
        Graph[a].push_back(b);



        E--;
    }


    // fill(visited, visited+V, 0);

    int order[V];
    int pos = V-1;

    for(int i = 0; i < V; i++){
        if(!visited[i]){
            vector<int> visitedByDFS;
            dfs(i, visitedByDFS, visited, V, Graph);
            for(int j = 0, len = visitedByDFS.size(); j<len; j++){
                order[pos] = visitedByDFS[j];
                pos--;
            }
        }
    }


    for(int node : order){
        cout << node << " ";
    }





}

